blog/_src/posts/2017-09-27-tutorial-contributing-to-racket.md

```diff
@@ -0,0 +1,414 @@
+
+
+## 2. Find the code
+
+A contribution typically needs an implementation, unit tests, and documentation.
+Finding the right files to edit to make these changes can be difficult,
+ but some general advice applies.
+
+
+### Finding the implementation
+
+If you want to changes a particular identifier: (1) open DrRacket, (2) import
+ the identifier, (3) right-click it, and (4) click "Open Defining File".
+The Emacs [`racket-mode`](https://github.com/greghendershott/racket-mode)
+ implements a similar shortcut.
+When it works, this is the fastest way to navigate a new codebase.
+
+If you know the name of the module you want to change, then DrRacket
+ or `racket-mode` can help jump to it.
+Sometimes you will find that the "public" module only reprovides identifiers
+ from a "private" one.
+Just keep following the module paths.
+
+A command-line alternative is [`raco-find-collection`](https://github.com/takikawa/raco-find-collection).
+Install it and run `raco fc <PKG>`.
+This prints the path to the `<PKG>` collection.
+Use this path to interpret Racket module paths.
+
+> For example, the module path `<PKG>` refers to the file `<PKG>/main.rkt`.
+> The module path `<PKG>/foo/bar/baz` refers to the file `<PKG>/foo/bar/baz.rkt`.
+
+
+### Finding the tests
+
+Most large packages have a directory named `<PKG>-test/` containing unit tests.
+The tests you need to edit should be _somewhere_ in here.
+
+Tests for the `racket/racket` repo may be under the:
+  [`racket-test-core/`](https://github.com/racket/racket/tree/master/pkgs/racket-test-core),
+  [`racket-test/`](https://github.com/racket/racket/tree/master/pkgs/racket-test), or
+  [`racket-test-extra/`](https://github.com/racket/racket/tree/master/pkgs/racket-test-extra)
+  directories.
+Note that many of these tests are written using a custom test library.
+Thankfully there are plenty of nearby examples showing how to use this library.
+
+Implementation files may define a `test` submodule.
+Try grepping for `module.*test`
+
+> If you are not sure where to put new tests,
+>  then just add a comment to your pull request with a test program.
+> A repo manager can tell you where the test belongs.
+
+
+### Finding the docs
+
+Most large package repos have a directory named `<PKG>-doc/` that contains the
+ documentation.
+Look for a directory named `scribblings/` under the `-doc/` directory.
+
+Small packages usually have a top-level directory named `scribblings/` with documentation.
+At any rate, the package's `info.rkt` file should have a `scribbings` field
+ whose value contains a path to the documentation.
+
+Documentation for core Racket is in the [`RKT_DIR/pkgs/racket-doc/scribblings/`](https://github.com/racket/racket/tree/master/pkgs/racket-doc/scribblings)
+ directory.
+In particular, the Racket Reference starts [here](https://github.com/racket/racket/blob/master/pkgs/racket-doc/scribblings/reference/reference.scrbl).
+
+If you can find the HTML version of the documentation using `my-raco docs`,
```

Comment:
Also: in the HTML docs, if you click on a heading, it will reveal the path to the underlying Scribble source.